src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+        _refreshToken = authenticatedResponse.at(L"refresh_token").as_string();
+
+        //step 3: getting the tenants and the required tenant id
+        json::value tenantsResponse = _getTenants();
+        json::array tenantList = tenantsResponse.at(L"value").as_array();
+
+        if (tenantList.size() == 0)
+        {
+            _outputHandlers(winrt::to_hstring("Could not find any tenants.\r\n"));
+            return -1;
+        }
+        else if (tenantList.size() == 1)
+        {
+            _tenantID = tenantList.at(0).at(L"tenantId").as_string();
+        }
+        else
+        {
+            _maxSize = tenantList.size();
+            for (int i = 0; i < _maxSize; i++)
+            {
+                utility::string_t str = L"Tenant " + std::to_wstring(i) + L": " + tenantList.at(i).at(L"displayName").as_string() + L" (" + tenantList.at(i).at(L"tenantId").as_string() + L")\r\n";
+                _outputHandlers(str);
+            }
+            utility::string_t str = L"Please enter the desired tenant number (a number from 0 to " + std::to_wstring((_maxSize - 1)) + L" inclusive).\r\n";
+            _outputHandlers(str);
+            _needInput = true;
+            // Use a lock to wait for the user to input a valid number
+            std::unique_lock<std::mutex> tenantNumberLock{ _tenantNumMutex };
+            _canProceed.wait(tenantNumberLock, [=]() {
+                return (_tenantNumber >= 0 && _tenantNumber < _maxSize) || _closing.load();
+            });
+            // User might have closed the tab while we waited for input
+            if (_closing.load())
+            {
+                return -1;
+            }
+            _needInput = false;
+            _tenantID = tenantList.at(_tenantNumber).at(L"tenantId").as_string();
+        }
+
+        //step 4: refresh token to get a new one specific to the desired tenant ID
+        json::value refreshResponse = _refreshTokens();
+        _accessToken = refreshResponse.at(L"access_token").as_string();
+        _refreshToken = refreshResponse.at(L"refresh_token").as_string();
+
+        //step 5: use new access token to get user's cloud shell settings
+        json::value settingsResponse = _getCloudShellUserSettings();
+        if (settingsResponse.has_field(L"error"))
+        {
+            _outputHandlers(winrt::to_hstring("You have not set up your cloud shell account yet. Please go to https://shell.azure.com to set it up.\r\n"));
+            return -1;
+        }
+
+        //step 6: request for a cloud shell
+        _outputHandlers(winrt::to_hstring("Requesting for a cloud shell..."));
+        _cloudShellUri = _getCloudShell();
+        _outputHandlers(winrt::to_hstring("Succeeded.\r\n"));
+
+        //step 7: request for a terminal for said cloud shell
+        //We only support bash for now, so don't bother with the user's preferred shell
+        //fyi: we can't call powershell yet because it sends VT sequences we don't support yet

```

Comment:
When we leave things behind like this in comments, the policy is to create a new issue that represents the outstanding work and then reference it here as `// TODO: GitHub #<number> - we can't call powershell yet...`